### PR TITLE
Redirect to community channel when following universal link

### LINKED ISSUE #19527: Redirect to community channel when following universal link
# Feature Issue

### User Story

As a user, I want to be redirected to a community channel when following a channel universal link instead of being redirected to the channel list. 

### Description

Currently, the user is redirected to the community overview screen with the channel list when they follow a channel link.

### Designs 

https://www.figma.com/file/q3kZPl7OpbB7btFK7HG0Mr/Sharing-for-Mobile?type=design&node-id=2344-285019&mode=design&t=jGlRlWtmiBU8VQQT-4

### Acceptance Criteria

User is redirected to the channel when following a channel link, e.g. `https://status.app/cc/G4UAAGRkmyOZJaKId0bp6vd8pxw4tAWaBZYEFAgf7zHWt8nhSpA_gIDPEl4ZPkZuMlDEw3PsuObYttc-PX8CKEZh4hWloSCE19DNwI_tOXIQC2ifAilM8j70ms7ukzqyoyzp7KpPPLvOuizy07jv6h4=#zQ3shh5a8qt6oAJoR3tPHwCm1PwPGHgzhBegy3P5B69z1ZfGu`

### PR DESCRIPTION
**Pull request by Parveshdhull**
fixes https://github.com/status-im/status-mobile/issues/19527

### Testing
Please make sure universal links for all types are working as expected

status: ready

### CODE REVIEW

File: src/status_im/common/shared_urls/data_store.cljs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+(ns status-im.common.shared-urls.data-store
+  (:require
+    [clojure.set :as set]))
+
+(defn <-rpc
+  [response]
```
Comment (ibrkhalil) [inline src/status_im/common/shared_urls/data_store.cljs]: We can destructure community, channel, and contact from response.

**Comment by yqrashawn** [comment]
I recall there's a decision made not to parse encoded data in universal links on mobile app
and for community channel links, redirect them to community overview because the channel info is in the encoded data 

[related code](https://github.com/status-im/status-mobile/blob/06aa4671b69f563e3edf20516173ade6a2975127/src/status_im/common/router.cljs#L103) that redirects /cc/ link to community overview
note the code below for `(= handler :community-chat)` won't work for universal links, it's for [deep link in notification](https://github.com/status-im/status-go/blob/c921079761cf15c835901eeca254233e2ed79f5f/protocol/chat.go#L535)

Link to the decision https://www.notion.so/Usage-of-link-data-7cf6c9e2a9554761baae9a1ba84e70c2
I don't know if things changed now. cc @cammellos

**Comment by Parveshdhull** [comment]
Thank you @yqrashawn for sharing above discussion.

> I don't know if things changed now

Confirmed with Andrea, We should not use encoded data in order to display information. But its fine to redirect to the channel.

**Review by ibrkhalil — APPROVED** [review]

**Comment by yqrashawn** [comment]
Thanks for the updates @Parveshdhull 

I suggest following the old path to add conds into `match-uri` and `handle-uri` 
and maybe handle things in `match-community-channel-in-encoded-data-async` as below diff and then redirect the logic to shared-urls

```diff
diff --git a/src/status_im/common/router.cljs b/src/status_im/common/router.cljs
index 08d7f4d97b..9ec58d4e74 100644
--- a/src/status_im/common/router.cljs
+++ b/src/status_im/common/router.cljs
@@ -100,7 +100,7 @@
       ens?
       (assoc-in [:route-params :ens-name] fragment)
 
-      (and (or (= handler :community) (= handler :community-chat)) compressed-key?)
+      (and (= handler :community) compressed-key?)
       (assoc-in [:route-params :community-id] fragment)
 
       (and equal-end-of-base64url (= handler :community) (:community-data route-params))
@@ -119,6 +119,11 @@
            (re-find constants/regx-starts-with-uuid (:community-data route-params)))
       (assoc-in [:route-params :community-channel-id] (:community-data route-params))
 
+      (and fragment
+           (= handler :community-chat)
+           (:community-data route-params))
+      (assoc-in [:route-params :community-channel-id-in-encoded-data] (:community-data route-params))
+
       (and equal-end-of-base64url (= handler :user) (:user-data route-params))
       (update-in [:route-params :user-data] #(str % equal-end-of-base64url))
 
@@ -212,6 +217,10 @@
     (cb {:type  :community-chat
          :error :not-found})))
 
+(defn match-community-channel-in-encoded-data-async
+  [{:keys [community-channel-id-in-encoded-data community-id]} cb]
+)
+
 (defn match-browser
   [uri {:keys [domain]}]
   ;; NOTE: We rebuild domain from original URI and matched domain
@@ -307,6 +316,11 @@
       (and (= handler :community-chat) (:community-channel-id route-params) (:community-id route-params))
       (match-community-channel-async route-params cb)
 
+      (and (= handler :community-chat)
+           (:community-channel-id-in-encoded-data route-params)
+           (:community-id route-params))
+      (match-community-channel-in-encoded-data-async route-params cb)
+
       (and (= handler :community-chat) (:community-id route-params))
       (cb {:type         :community
            :community-id (:community-id route-params)})
```

**Comment by Parveshdhull** [comment]
Thank you very much @yqrashawn for your suggestions.

As we discussed lets keep current solution for now, because due to mix of deep links and shared links things are little messed up and these files probably need some refactoring later.

**Review by yqrashawn — APPROVED** [review]

**Comment by pavloburykh** [comment]
@Parveshdhull thanks for the PR! Please take a look at the issues.

### ISSUE 1 App crashes when following community/channel link from external browser


**Steps:** 

1. Open community/channel link in external browser
2. Wait until web page is loaded
3. Tap 'Join community in Status' 
4. Observe the result

**Expected result:** community/channel screen is opened in Status app. 

**Actual result:** app crashes after redirection. 

[Status-debug-logs - 2024-04-16T122944.531.zip](https://github.com/status-im/status-mobile/files/14994790/Status-debug-logs.-.2024-04-16T122944.531.zip)

https://github.com/status-im/status-mobile/assets/97245802/c771ca8a-6804-4db9-aef6-9113a727b7be

**Comment by pavloburykh** [comment]
### ISSUE 2 User is able to access locked token gated channel by following the link

**Steps:** 

1. Follow the link to token gated channel which is locked for the user
2. Observe the result

**Actual result:** token gated channel screen is opened. 


https://github.com/status-im/status-mobile/assets/97245802/cf02f2fe-3afc-4fef-906b-d1b8164ee252



**Expected result:** @Parveshdhull I am not sure what should be the expected result. By default user is not able to access locked channels. In case of following the link - he is able to access. Though he will not see channel history and is unable to post messages. So the issue seems to be minor but we still need to understand how should we handle this case. BTW, the same workaround exists by following hashtag links.

**Comment by Parveshdhull** [comment]
hi @pavloburykh, Thank you very much for finding these issues.

Issue 1 should be fixed now. 

For Issue 2 we need to figure out what should be expected behavior. Should we just navigate to community screen?
Also, as you mentioned Issue 2 is also reproducible by following hashtag's, will it be ok if we fix it separately?

**Comment by pavloburykh** [comment]
@Parveshdhull thank you! Will check fix for issue 1. 

> For Issue 2 we need to figure out what should be expected behavior. Should we just navigate to community screen?
Also, as you mentioned Issue 2 is also reproducible by following hashtag's, will it be ok if we fix it separately?

Agree, I will log it separately and ping design team for input.

**Comment by Parveshdhull** [comment]
> Agree, I will log it separately and ping design team for input.

Thank you

Please hold testing for a moment, it seems tests failed. Will push changes again to fix them.

**Comment by Parveshdhull** [comment]
> Please hold testing for a moment, it seems tests failed. Will push changes again to fix them.

fixed

**Comment by pavloburykh** [comment]
@Parveshdhull please, take a look at the issue below. Probably we will have to handle it separately as high likely it will require additional changes on go side, but tell WDYT.

### ISSUE 3 Channel will never be fetched if community has not been opened and fetched before

**Example of channel link** https://status.app/cc/G5EAAGRqm6gnoibf_r1zxIUu4qUMHHLg8C3BwAKxjEIKPGT9_ZFqYv1FS3-W8CFKY9QmgyJ-HptlMj8fj8_T9vHN9BxBMuqHEljBZPm2JRknkTf_zD8_dr_PPJYyPe8HBBEuZUGUQDElS4gZB82YTkGDKATNFLFFgwBermYLNMeEAA==#zQ3shWQuxzW3NsSxSA4VExZjLKDRMUwLDuEXdfcRHbQ1aDjeC

**Example of community link which above channel belongs to** https://status.app/c/ixmACjIKEE5ldyBjb21tdW5pdHkgMSASDmN4Y3h6Y3h6Y3h6enhjGAIiByNERjQzNDUqAwABAgM=#zQ3shWQuxzW3NsSxSA4VExZjLKDRMUwLDuEXdfcRHbQ1aDjeC

**Preconditions:** let's imagine someone shared you some channel link which belongs to community which has not been opened and fetched before on your account. 

**Steps:** 

1. Open channel link
2. See if channel will be fetched

**Expected result:** channel is fetched

**Actual result:** channel will never be fetched until user opens and fetches community to which this channel belongs. 


https://github.com/status-im/status-mobile/assets/97245802/1ce775a2-aac0-48d2-b958-ebee700f814b

**Comment by Parveshdhull** [comment]
thank you finding issue 3, should be fixed now

**Comment by pavloburykh** [comment]
@Parveshdhull thanks a lot for your fixes. PR is ready for merge.

**Pull request merged**